flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/SlotPoolImpl.java

```diff
@@ -648,26 +648,8 @@ boolean offerSlot(
 			slotOffer.getResourceProfile(),
 			taskManagerGateway);
 
-		// check whether we have request waiting for this slot
-		PendingRequest pendingRequest = pendingRequests.removeKeyB(allocationID);
```

Comment:
I am not sure about all consequences of this change for the existing scheduling. I mean that we do not respect SlotRequestId->AllocationID by accepting the slot offer. Would it make sense to keep this behaviour configurable for now depending on scheduling strategy? Or this complication is not needed?

Reply:
I think it's a good idea to make it configurable. 

Besides the benefit to reduce risk for streaming and DataSet jobs, another benefit is that we can also drop the change to remap orphaned allocations. This is because the remapping is for fail-fast of pending requests in `failAllocation(...)` which makes difference only if it is a streaming job.

Actually I'm thinking whether we need to keep the fail-fast mechanism in `failAllocation(...)` in the future. It requires the slot pool to differentiate streaming requests and batch requests. And in the future if a slotpool contains both batch slots(occupied temporarily) and streaming slots(occupied indefinitely), a failed allocation for streaming request does not need to fail immediately if it is still fulfillable, just like how we currently deal with batch requests.

What do you think of dropping the commit to remap orphaned allocations?

Reply:
There is this `UnfulfillableSlotRequestException` which is still a fail fast route if RM finds that a certain request profile cannot be fulfilled at all with any existing slot and cannot be allocated. It is relevant for both batch and streaming and bulk as I see. I do not know the whole background of this. At first glance, it seems to be a complication but it might be necessary.

Reply:
I had another thought and prefers to not make it configurable to fulfill request in request order.
The FIFO order works for any scheduling strategy and is even a improvement for lazy from source scheduling. There was a ticket for the same purpose "[FLINK-13165] Complete slot requests in request order" although it did not fully make it at last. `LinkedHashMap` was introduced by it and eases this PR. `SlotPoolRequestCompletionTest` was also introduced by it and can be reused.
This PR is more alike a fix to FLINK-13165.

The orphaned allocation remapping also ensures that the fail-fast route will not break with this change.

What do you think?

Reply:
Ok, the FLINK-13165 makes slot requests to be completed in order only if the offers come with unknown `AllocationIds`, right? Generally we expect that RM keeps the `AllocationId` to match `SlotRequestID`. I am fine to break the tie `SlotRequestId->AllocationID` if there is no known consequences. Eventually, I hope it might even help to simplify the `SlotPoolImpl`.

Reply:
I think it should be fine to remap `SlotRequestIds` and `AllocationIds`. In the end it is an implementation detail of the `SlotPoolImpl` how a `SlotRequestId` is mapped to an `AllocationId` and no user of the `SlotPool` should rely on/use it. The only important bit is to assign the orphaned `AllocationId` so that we can react to signals from the `ResourceManager`.

Reply:
Thanks for confirming it.